llm/ggml_llama.go

```diff
@@ -0,0 +1,672 @@
+		TypicalP:         llm.TypicalP,
+		RepeatLastN:      llm.RepeatLastN,
+		RepeatPenalty:    llm.RepeatPenalty,
+		PresencePenalty:  llm.PresencePenalty,
+		FrequencyPenalty: llm.FrequencyPenalty,
+		Mirostat:         llm.Mirostat,
+		MirostatTau:      llm.MirostatTau,
+		MirostatEta:      llm.MirostatEta,
+		PenalizeNl:       llm.PenalizeNewline,
+		Stop:             llm.Stop,
+	}
+	data, err := json.Marshal(predReq)
+	if err != nil {
+		return fmt.Errorf("error marshaling data: %v", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewBuffer(data))
+	if err != nil {
+		return fmt.Errorf("error creating POST request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("POST predict: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 400 {
+		bodyBytes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("failed reading llm error response: %w", err)
+		}
+		log.Printf("llm predict error: %s", bodyBytes)
+		return fmt.Errorf("%s", bodyBytes)
+	}
+
+	reader := bufio.NewReader(resp.Body)
+	var genCtx strings.Builder
+	genCtx.WriteString(trimmedPrompt)
+	if err != nil {
+		return fmt.Errorf("decode prompt context: %v", err)
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			// This handles the request cancellation
+			return ctx.Err()
+		default:
+			line, err := reader.ReadString('\n')
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return fmt.Errorf("error reading llm response: %v", err)
+			}
+			if line == "\n" {
+				continue
+			}
+
+			// Read data from the server-side event stream
+			if len(line) > 6 && line[:6] == "data: " {
+				evt := line[6:]
+				var complete PredictComplete
+				if err := json.Unmarshal([]byte(evt), &complete); err != nil {
+					return fmt.Errorf("error unmarshaling llm complete response: %v", err)
+				}
+
+				if complete.Timings.PredictedMS > 0 {
+					genCtx.WriteString(complete.Content)
+					embd, err := llm.Encode(ctx, genCtx.String())
+					if err != nil {
+						return fmt.Errorf("encoding context: %v", err)
+					}
+					fn(api.GenerateResponse{
+						Done:               true,
+						Context:            embd,
+						PromptEvalCount:    int(complete.Timings.PromptN),
+						PromptEvalDuration: parseDurationMs(float64(complete.Timings.PromptMS)),
+						EvalCount:          int(complete.Timings.PredictedN),
+						EvalDuration:       parseDurationMs(float64(complete.Timings.PredictedMS)),
+					})
+					return nil
+				}
+
+				var pred Prediction
+				if err := json.Unmarshal([]byte(evt), &pred); err != nil {
+					return fmt.Errorf("error unmarshaling llm prediction response: %v", err)
+				}
+				genCtx.WriteString(pred.Content)
+				fn(api.GenerateResponse{Response: pred.Content})
+			}
+		}
+	}
+}
+
+func (llm *llama) marshalPrompt(ctx context.Context, pCtx []int, prompt string) (string, error) {
```

Comment:
this isn't quite right, working on fixing it now

Reply:
This can be passed directly to the server which will handle truncating